// @glittr-sdk/sdk/dist/transaction/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = void 0;
exports.schema = {
    struct: {
        transfer: {
            option: {
                struct: {
                    transfers: {
                        array: {
                            type: {
                                struct: {
                                    asset: { array: { type: { array: { type: 'u8' } }, len: 2 } },
                                    output: { array: { type: 'u8' } },
                                    amount: { array: { type: 'u8' } }, // VARUINT
                                }
                            },
                            showLen: true
                        }
                    }
                }
            }
        },
        contract_creation: {
            option: {
                struct: {
                    contract_type: {
                        enum: [
                            {
                                struct: {
                                    moa: {
                                        struct: {
                                            ticker: {
                                                option: {
                                                    struct: {
                                                        number: { array: { type: 'u8' } },
                                                        spacers: { option: { array: { type: 'u8' } } }
                                                    }
                                                }
                                            },
                                            supply_cap: { option: { array: { type: 'u8' } } },
                                            divisibility: 'u8',
                                            live_time: { array: { type: 'u8' } },
                                            end_time: { option: { array: { type: 'u8' } } },
                                            mint_mechanism: {
                                                struct: {
                                                    preallocated: {
                                                        option: {
                                                            struct: {
                                                                allocations: {
                                                                    map: {
                                                                        key: { array: { type: 'u8' } },
                                                                        value: {
                                                                            enum: [
                                                                                {
                                                                                    struct: { vec_pubkey: { array: { type: { array: { type: 'u8', showLen: true } }, showLen: true } } },
                                                                                },
                                                                                {
                                                                                    struct: {
                                                                                        bloom_filter: {
                                                                                            struct: {
                                                                                                filter: { array: { type: 'u8' } },
                                                                                                arg: {
                                                                                                    enum: [{ struct: { tx_id: { struct: {} } } }]
                                                                                                }
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            ]
                                                                        }
                                                                    }
                                                                },
                                                                vesting_plan: {
                                                                    option: {
                                                                        enum: [
                                                                            { struct: { timelock: { array: { type: 'u8' } } } },
                                                                            {
                                                                                struct: {
                                                                                    scheduled: {
                                                                                        array: {
                                                                                            type: {
                                                                                                struct: {
                                                                                                    ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } },
                                                                                                    tolerance: 'i64'
                                                                                                }
                                                                                            },
                                                                                            showLen: true
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        ]
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    },
                                                    free_mint: {
                                                        option: {
                                                            struct: {
                                                                supply_cap: { option: { array: { type: 'u8' } } },
                                                                amount_per_mint: { array: { type: 'u8' } }, // VARUINT
                                                            }
                                                        }
                                                    },
                                                    purchase: {
                                                        option: {
                                                            struct: {
                                                                input_asset: {
                                                                    enum: [
                                                                        { struct: { raw_btc: { struct: {} } } },
                                                                        { struct: { glittr_asset: { array: { type: { array: { type: 'u8' } }, len: 2 } } } },
                                                                        { struct: { rune: { struct: {} } } },
                                                                        { struct: { ordinal: { struct: {} } } }
                                                                    ]
                                                                },
                                                                pay_to_key: { option: { array: { type: 'u8' } } },
                                                                ratio: {
                                                                    enum: [
                                                                        {
                                                                            struct: {
                                                                                fixed: {
                                                                                    struct: {
                                                                                        ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                    }
                                                                                }
                                                                            }
                                                                        },
                                                                        {
                                                                            struct: {
                                                                                oracle: {
                                                                                    struct: {
                                                                                        setting: {
                                                                                            struct: {
                                                                                                pubkey: { array: { type: 'u8' } },
                                                                                                asset_id: { option: 'string' },
                                                                                                block_height_slippage: 'u8'
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    ]
                                                                }
                                                            }
                                                        }
                                                    },
                                                }
                                            },
                                            commitment: {
                                                option: {
                                                    struct: {
                                                        public_key: { array: { type: 'u8' } },
                                                        args: {
                                                            struct: {
                                                                fixed_string: {
                                                                    option: {
                                                                        struct: {
                                                                            number: { array: { type: 'u8' } },
                                                                            spacers: { option: { array: { type: 'u8' } } }
                                                                        }
                                                                    }
                                                                },
                                                                string: 'string'
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    mba: {
                                        struct: {
                                            ticker: {
                                                option: {
                                                    struct: {
                                                        number: { array: { type: 'u8' } },
                                                        spacers: { option: { array: { type: 'u8' } } }
                                                    }
                                                }
                                            },
                                            supply_cap: { option: { array: { type: 'u8' } } },
                                            divisibility: 'u8',
                                            live_time: { array: { type: 'u8' } },
                                            end_time: { option: { array: { type: 'u8' } } },
                                            mint_mechanism: {
                                                struct: {
                                                    preallocated: {
                                                        option: {
                                                            struct: {
                                                                allocations: {
                                                                    map: {
                                                                        key: { array: { type: 'u8' } },
                                                                        value: {
                                                                            enum: [
                                                                                {
                                                                                    struct: { vec_pubkey: { array: { type: { array: { type: 'u8', showLen: true } }, showLen: true } } },
                                                                                },
                                                                                {
                                                                                    struct: {
                                                                                        bloom_filter: {
                                                                                            struct: {
                                                                                                filter: { array: { type: 'u8' } },
                                                                                                arg: {
                                                                                                    enum: [{ struct: { tx_id: { struct: {} } } }]
                                                                                                }
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            ]
                                                                        }
                                                                    }
                                                                },
                                                                vesting_plan: {
                                                                    option: {
                                                                        enum: [
                                                                            { struct: { timelock: { array: { type: 'u8' } } } },
                                                                            {
                                                                                struct: {
                                                                                    scheduled: {
                                                                                        array: {
                                                                                            type: {
                                                                                                struct: {
                                                                                                    ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } },
                                                                                                    tolerance: 'i64'
                                                                                                }
                                                                                            },
                                                                                            showLen: true
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        ]
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    },
                                                    free_mint: {
                                                        option: {
                                                            struct: {
                                                                supply_cap: { option: { array: { type: 'u8' } } },
                                                                amount_per_mint: { array: { type: 'u8' } }, // VARUINT
                                                            }
                                                        }
                                                    },
                                                    purchase: {
                                                        option: {
                                                            struct: {
                                                                input_asset: {
                                                                    enum: [
                                                                        { struct: { raw_btc: { struct: {} } } },
                                                                        { struct: { glittr_asset: { array: { type: { array: { type: 'u8' } }, len: 2 } } } },
                                                                        { struct: { rune: { struct: {} } } },
                                                                        { struct: { ordinal: { struct: {} } } }
                                                                    ]
                                                                },
                                                                pay_to_key: { option: { array: { type: 'u8' } } },
                                                                ratio: {
                                                                    enum: [
                                                                        {
                                                                            struct: {
                                                                                fixed: {
                                                                                    struct: {
                                                                                        ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                    }
                                                                                }
                                                                            }
                                                                        },
                                                                        {
                                                                            struct: {
                                                                                oracle: {
                                                                                    struct: {
                                                                                        setting: {
                                                                                            struct: {
                                                                                                pubkey: { array: { type: 'u8' } },
                                                                                                asset_id: { option: 'string' },
                                                                                                block_height_slippage: 'u8'
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    ]
                                                                }
                                                            }
                                                        }
                                                    },
                                                    collateralized: {
                                                        option: {
                                                            struct: {
                                                                input_asset: {
                                                                    enum: [
                                                                        { struct: { raw_btc: { struct: {} } } },
                                                                        { struct: { glittr_asset: { array: { type: { array: { type: 'u8' } }, len: 2 } } } },
                                                                        { struct: { rune: { struct: {} } } },
                                                                        { struct: { ordinal: { struct: {} } } }
                                                                    ]
                                                                },
                                                                _mutable_assets: 'bool',
                                                                mint_structure: {
                                                                    enum: [
                                                                        {
                                                                            struct: {
                                                                                ratio: {
                                                                                    enum: [
                                                                                        {
                                                                                            struct: {
                                                                                                fixed: {
                                                                                                    struct: {
                                                                                                        ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                                    }
                                                                                                }
                                                                                            }
                                                                                        },
                                                                                        {
                                                                                            struct: {
                                                                                                oracle: {
                                                                                                    struct: {
                                                                                                        setting: {
                                                                                                            struct: {
                                                                                                                pubkey: { array: { type: 'u8' } },
                                                                                                                asset_id: { option: 'string' },
                                                                                                                block_height_slippage: 'u8'
                                                                                                            }
                                                                                                        }
                                                                                                    }
                                                                                                }
                                                                                            }
                                                                                        }
                                                                                    ]
                                                                                }
                                                                            }
                                                                        },
                                                                        {
                                                                            struct: {
                                                                                proportional: {
                                                                                    struct: {
                                                                                        ratio_model: { enum: [{ struct: { constant_product: { struct: {} } } }] },
                                                                                        inital_mint_pointer_to_key: { option: { array: { type: 'u8' } } }, //VARUINT
                                                                                    }
                                                                                }
                                                                            }
                                                                        },
                                                                        {
                                                                            struct: {
                                                                                account: {
                                                                                    struct: {
                                                                                        max_ltv: { array: { type: { array: { type: 'u8' } }, len: 2 } },
                                                                                        ratio: {
                                                                                            enum: [
                                                                                                {
                                                                                                    struct: {
                                                                                                        fixed: {
                                                                                                            struct: {
                                                                                                                ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                                            }
                                                                                                        }
                                                                                                    }
                                                                                                },
                                                                                                {
                                                                                                    struct: {
                                                                                                        oracle: {
                                                                                                            struct: {
                                                                                                                setting: {
                                                                                                                    struct: {
                                                                                                                        pubkey: { array: { type: 'u8' } },
                                                                                                                        asset_id: { option: 'string' },
                                                                                                                        block_height_slippage: 'u8'
                                                                                                                    }
                                                                                                                }
                                                                                                            }
                                                                                                        }
                                                                                                    }
                                                                                                }
                                                                                            ]
                                                                                        }
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    ]
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            burn_mechanism: {
                                                struct: {
                                                    return_collateral: {
                                                        option: {
                                                            struct: {
                                                                fee: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                                                                oracle_setting: {
                                                                    option: {
                                                                        struct: {
                                                                            pubkey: { array: { type: 'u8' } },
                                                                            asset_id: { option: 'string' },
                                                                            block_height_slippage: 'u8'
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            swap_mechanism: {
                                                struct: {
                                                    fee: { option: { array: { type: 'u8' } } }, // VARUINT
                                                }
                                            },
                                            commitment: {
                                                option: {
                                                    struct: {
                                                        public_key: { array: { type: 'u8' } },
                                                        args: {
                                                            struct: {
                                                                fixed_string: {
                                                                    option: {
                                                                        struct: {
                                                                            number: { array: { type: 'u8' } },
                                                                            spacers: { option: { array: { type: 'u8' } } }
                                                                        }
                                                                    }
                                                                },
                                                                string: 'string'
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    spec: {
                                        struct: {
                                            spec: {
                                                enum: [
                                                    {
                                                        struct: {
                                                            mint_only_asset: {
                                                                struct: {
                                                                    input_asset: {
                                                                        option: {
                                                                            enum: [
                                                                                { struct: { raw_btc: { struct: {} } } },
                                                                                { struct: { glittr_asset: { array: { type: { array: { type: 'u8' } }, len: 2 } } } },
                                                                                { struct: { rune: { struct: {} } } },
                                                                                { struct: { ordinal: { struct: {} } } }
                                                                            ]
                                                                        },
                                                                    },
                                                                    peg_in_type: {
                                                                        option: {
                                                                            enum: [
                                                                                { struct: { pubkey: { array: { type: 'u8' } } } },
                                                                                { struct: { burn: { struct: {} } } }
                                                                            ]
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    },
                                                    {
                                                        struct: {
                                                            mint_burn_asset: {
                                                                struct: {
                                                                    collateralized: {
                                                                        option: {
                                                                            struct: {
                                                                                _mutable_assets: 'bool',
                                                                                input_asset: {
                                                                                    option: {
                                                                                        array: {
                                                                                            type: {
                                                                                                enum: [
                                                                                                    { struct: { raw_btc: { struct: {} } } },
                                                                                                    { struct: { glittr_asset: { array: { type: { array: { type: 'u8' } }, len: 2 } } } },
                                                                                                    { struct: { rune: { struct: {} } } },
                                                                                                    { struct: { ordinal: { struct: {} } } }
                                                                                                ]
                                                                                            }
                                                                                        }
                                                                                    }
                                                                                },
                                                                                mint_structure: {
                                                                                    option: {
                                                                                        enum: [
                                                                                            {
                                                                                                struct: {
                                                                                                    ratio: {
                                                                                                        enum: [
                                                                                                            {
                                                                                                                struct: {
                                                                                                                    fixed: {
                                                                                                                        struct: {
                                                                                                                            ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                                                        }
                                                                                                                    }
                                                                                                                }
                                                                                                            },
                                                                                                            {
                                                                                                                struct: {
                                                                                                                    oracle: {
                                                                                                                        struct: {
                                                                                                                            setting: {
                                                                                                                                struct: {
                                                                                                                                    pubkey: { array: { type: 'u8' } },
                                                                                                                                    asset_id: { option: 'string' },
                                                                                                                                    block_height_slippage: 'u8'
                                                                                                                                }
                                                                                                                            }
                                                                                                                        }
                                                                                                                    }
                                                                                                                }
                                                                                                            }
                                                                                                        ]
                                                                                                    }
                                                                                                }
                                                                                            },
                                                                                            {
                                                                                                struct: {
                                                                                                    proportional: {
                                                                                                        struct: {
                                                                                                            ratio_model: { enum: [{ struct: { constant_product: { struct: {} } } }] },
                                                                                                            inital_mint_pointer_to_key: { option: { array: { type: 'u8' } } }, //VARUINT
                                                                                                        }
                                                                                                    }
                                                                                                }
                                                                                            },
                                                                                            {
                                                                                                struct: {
                                                                                                    account: {
                                                                                                        struct: {
                                                                                                            max_ltv: { array: { type: { array: { type: 'u8' } }, len: 2 } },
                                                                                                            ratio: {
                                                                                                                enum: [
                                                                                                                    {
                                                                                                                        struct: {
                                                                                                                            fixed: {
                                                                                                                                struct: {
                                                                                                                                    ratio: { array: { type: { array: { type: 'u8' } }, len: 2 } } // VARUINT Fraction
                                                                                                                                }
                                                                                                                            }
                                                                                                                        }
                                                                                                                    },
                                                                                                                    {
                                                                                                                        struct: {
                                                                                                                            oracle: {
                                                                                                                                struct: {
                                                                                                                                    setting: {
                                                                                                                                        struct: {
                                                                                                                                            pubkey: { array: { type: 'u8' } },
                                                                                                                                            asset_id: { option: 'string' },
                                                                                                                                            block_height_slippage: 'u8'
                                                                                                                                        }
                                                                                                                                    }
                                                                                                                                }
                                                                                                                            }
                                                                                                                        }
                                                                                                                    }
                                                                                                                ]
                                                                                                            }
                                                                                                        }
                                                                                                    }
                                                                                                }
                                                                                            }
                                                                                        ]
                                                                                    }
                                                                                }
                                                                            }
                                                                        }
                                                                    }
                                                                }
                                                            }
                                                        }
                                                    },
                                                ]
                                            },
                                            pointer: { option: { array: { type: 'u8' } } },
                                            block_tx: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } } //VARUINT BlockTxTuple
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    nft: {
                                        struct: {
                                            asset: { array: { type: 'u8', showLen: true } },
                                            supply_cap: { option: { array: { type: 'u8' } } },
                                            live_time: { array: { type: 'u8' } },
                                            end_time: { option: { array: { type: 'u8' } } },
                                            pointer: { option: { array: { type: 'u8' } } }, // VARUINT
                                        }
                                    }
                                }
                            }
                        ]
                    },
                    spec: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } }, //VARUINT BlockTxTuple
                }
            }
        },
        contract_call: {
            option: {
                struct: {
                    contract: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                    call_type: {
                        enum: [
                            {
                                struct: {
                                    mint: {
                                        struct: {
                                            pointer: { option: { array: { type: 'u8' } } },
                                            oracle_message: {
                                                option: {
                                                    struct: {
                                                        signature: { array: { type: 'u8' } },
                                                        message: {
                                                            struct: {
                                                                input_outpoint: { option: { struct: { txid: 'string', vout: 'u32' } } },
                                                                min_in_value: { option: { array: { type: 'u8' } } },
                                                                out_value: { option: { array: { type: 'u8' } } },
                                                                asset_id: { option: 'string' },
                                                                ratio: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                                                                ltv: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                                                                outstanding: { option: { array: { type: 'u8' } } },
                                                                block_height: { array: { type: 'u8' } } // VARUINT
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            pointer_to_key: { option: { array: { type: 'u8' } } },
                                            assert_values: {
                                                option: {
                                                    struct: {
                                                        input_values: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        total_collateralized: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        min_out_value: { option: { array: { type: 'u8' } } }
                                                    }
                                                }
                                            },
                                            commitment_message: {
                                                option: {
                                                    struct: {
                                                        public_key: { array: { type: 'u8' } },
                                                        args: { array: { type: 'u8' } }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    burn: {
                                        struct: {
                                            pointer: { option: { array: { type: 'u8' } } },
                                            oracle_message: {
                                                option: {
                                                    struct: {
                                                        signature: { array: { type: 'u8' } },
                                                        message: {
                                                            struct: {
                                                                input_outpoint: { option: { struct: { txid: 'string', vout: 'u32' } } },
                                                                min_in_value: { option: { array: { type: 'u8' } } },
                                                                out_value: { option: { array: { type: 'u8' } } },
                                                                asset_id: { option: 'string' },
                                                                ratio: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                                                                ltv: { option: { array: { type: { array: { type: 'u8' } }, len: 2 } } },
                                                                outstanding: { option: { array: { type: 'u8' } } },
                                                                block_height: { array: { type: 'u8' } } // VARUINT
                                                            }
                                                        }
                                                    }
                                                }
                                            },
                                            pointer_to_key: { option: { array: { type: 'u8' } } },
                                            assert_values: {
                                                option: {
                                                    struct: {
                                                        input_values: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        total_collateralized: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        min_out_value: { option: { array: { type: 'u8' } } }
                                                    }
                                                }
                                            },
                                            commitment_message: {
                                                option: {
                                                    struct: {
                                                        public_key: { array: { type: 'u8' } },
                                                        args: { array: { type: 'u8' } }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    swap: {
                                        struct: {
                                            pointer: { array: { type: 'u8' } },
                                            assert_values: {
                                                option: {
                                                    struct: {
                                                        input_values: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        total_collateralized: { option: { array: { type: { array: { type: 'u8' } } } } },
                                                        min_out_value: { option: { array: { type: 'u8' } } }
                                                    }
                                                }
                                            },
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    open_account: {
                                        struct: {
                                            pointer_to_key: { array: { type: 'u8' } },
                                            share_amount: { array: { type: 'u8' } }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    close_account: {
                                        struct: {
                                            pointer: { array: { type: 'u8' } }
                                        }
                                    }
                                }
                            },
                            {
                                struct: {
                                    update_nft: {
                                        struct: {
                                            whitelist_address_bloom_filter: { option: { array: { type: 'u8', showLen: true } } },
                                            trusted_marketplace_fee_addresses: { option: { array: { type: 'string' } } },
                                            access_key_pointer: { option: { array: { type: 'u8' } } } // VARUINT
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            }
        }
    }
};
